exporter/instanaexporter/exporter.go

```diff
@@ -0,0 +1,144 @@
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/instanaexporter/internal/converter"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/instanaexporter/internal/converter/model"
+)
+
+type instanaExporter struct {
+	config    *Config
+	client    *http.Client
+	settings  component.TelemetrySettings
+	userAgent string
+}
+
+func (e *instanaExporter) start(_ context.Context, host component.Host) error {
+	client, err := e.config.HTTPClientSettings.ToClient(host, e.settings)
+	if err != nil {
+		return err
+	}
+	e.client = client
+	return nil
+}
+
+func (e *instanaExporter) pushConvertedTraces(ctx context.Context, td ptrace.Traces) error {
+	converter := converter.NewConvertAllConverter(e.settings.Logger)
+	spans := make([]model.Span, 0)
+
+	hostID := ""
+	resourceSpans := td.ResourceSpans()
+	for i := 0; i < resourceSpans.Len(); i++ {
+		resSpan := resourceSpans.At(i)
+
+		resource := resSpan.Resource()
+
+		hostIDAttr, ex := resource.Attributes().Get(backend.AttributeInstanaHostID)
+		if ex {
+			hostID = hostIDAttr.StringVal()
+		}
+
+		ilSpans := resSpan.ScopeSpans()
+		for j := 0; j < ilSpans.Len(); j++ {
+			converterBundle := converter.ConvertSpans(resource.Attributes(), ilSpans.At(j).Spans())
+
+			spans = append(spans, converterBundle.Spans...)
+		}
+	}
+
+	bundle := model.Bundle{Spans: spans}
+
+	if len(bundle.Spans) == 0 {
+		// skip exporting, nothing to do
+		return nil
+	}
+
+	req, err := bundle.Marshal()
+	if err != nil {
+		return consumererror.NewPermanent(err)
+	}
+
+	headers := map[string]string{
+		backend.HeaderKey:  e.config.AgentKey,
+		backend.HeaderHost: hostID,
+		backend.HeaderTime: "0",
+	}
+
+	return e.export(ctx, e.config.Endpoint, headers, req)
+}
+
+func newInstanaExporter(cfg config.Exporter, set component.ExporterCreateSettings) (*instanaExporter, error) {
+	iCfg := cfg.(*Config)
+
+	if iCfg.Endpoint != "" {
+		_, err := url.Parse(iCfg.Endpoint)
+		if err != nil {
+			return nil, errors.New("endpoint must be a valid URL")
+		}
+	}
+
+	userAgent := fmt.Sprintf("%s/%s (%s/%s)", set.BuildInfo.Description, set.BuildInfo.Version, runtime.GOOS, runtime.GOARCH)
+
+	return &instanaExporter{
+		config:    iCfg,
+		settings:  set.TelemetrySettings,
+		userAgent: userAgent,
+	}, nil
+}
+
+func (e *instanaExporter) export(ctx context.Context, url string, header map[string]string, request []byte) error {
+	url = strings.TrimSuffix(url, "/") + "/bundle"
+	e.settings.Logger.Debug("Preparing to make HTTP request", zap.String("url", url))
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(request))
+	if err != nil {
+		return consumererror.NewPermanent(err)
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("User-Agent", e.userAgent)
+
+	for name, value := range header {
+		req.Header.Set(name, value)
+	}
+
+	resp, err := e.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to make a HTTP request: %w", err)
```

Comment:
You could optimize this a bit: if this is a 4xx error type, it's a permanent error; if it's a 5xx error type, it's a regular error.